keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -9,105 +9,201 @@
 
-I'm proposing creating a new versioned grpc API under `k8s.io/kubernetes/pkg/serviceaccount`. This will be similar to how the KMS envelope encryption has an API at `k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/v1beta1/service.proto`
+### Notes/Constraints/Caveats
+
+- A new flag `--key-service-url` will be added to kube-apiserver specifying a unix socket where the key service will be accessible.
+- The flag `--key-service-url` can either be set to the location of a UDS on a filesystem, or be prefixed with an @ symbol and name a UDS in the abstract socket namespace.
+- combining local and external keys will not be supported.
+- Specifying both, `--service-account-key-file` and `--key-service-url` shall result in an error.
+
+### Risks and Mitigations
+
+- New token generation and validation could suffer a performance difference when reading over a socket, as an external process will be signing data.
+- Signing and verifying tokens over a grpc API carries the risk of a server side request forgery, where a malicious client could generate tokens. To mitigate this risk, the API will only be accessible over a unix socket.
+- External signers will only return signature and header while the token will be assembled by TokenGenerator implementation for external signing. This will help kube-apiserver control the payload and thus the claims.
+
+## Design Details
+
+### ExternalJWTSigner RPC
+
+Will be served on a socket as configured via the `--key-service-url` flag.
 
 ```proto
 syntax = "proto3";
 
-package v1alpha1;
-
-service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+// This service is served by a process on a local Unix Domain Socket.
+service ExternalJWTSigner {
+  // Sign takes a serialized JWT payload, and returns the serialized header and
+  // signature.  The caller can then assemble the JWT from the header, payload,
+  // and signature.
+  //
+  // The plugin MUST set a key id in the returned JWT header.
+  rpc Sign(SignJWTRequest) returns (SignJWTResponse) {}
+
+  // FetchKeys returns the set of public keys that are trusted to sign
+  // Kubernetes service account tokens. Kube-apiserver will call this RPC:
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
+message SignJWTRequest {
+  // The raw payload to be signed.  A serialized JSON object, no base64 wrapping.
```

Comment:
let's make this already be base64 encoded to remove a possible mistake in the external signer encoding (and symmetric with the "Already wrapped in URL-safe base64" fields in the response)

Reply:
Done